uaclient/security.py

```diff
@@ -465,69 +498,109 @@ def prompt_for_affected_packages(
+    if cloud_type in PRO_CLOUDS:
+        print(
+            status.MESSAGE_SECURITY_USE_PRO_TMPL.format(
+                title=CLOUD_TYPE_TO_TITLE.get(cloud_type), cloud=cloud_type
+            )
+        )
+    print(status.MESSAGE_SECURITY_UPDATE_NOT_INSTALLED_SUBSCRIPTION)
+    choice = util.prompt_choices(
+        "Choose: [S]ubscribe at ubuntu.com [A]ttach"
+        " existing token [C]ancel",
+        valid_choices=["s", "a", "c"],
+    )
+    if choice == "c":
+        return False
+    if choice == "a":
+        print(status.PROMPT_ENTER_TOKEN)
+        token = input("> ")
+        print(status.colorize_commands([["ua", "attach", token]]))
+        return bool(
+            0
+            == cli.action_attach(
+                argparse.Namespace(token=token, auto_enable=True), cfg
+            )
+        )
+    elif choice == "s":
+        print(status.PROMPT_UA_SUBSCRIPTION_URL)
+        print("TODO: GH: #1413: magic subscription attach")
+    return True
 
 
 def upgrade_packages_and_attach(
     cfg: UAConfig,
     upgrade_packages: "List[str]",
     upgrade_packages_ua: "List[str]",
-):
+) -> bool:
     """Upgrade available packages to fix a CVE.
 
     Upgrade all packages in Ubuntu standard updates regardless of attach
     status.
 
     For UA upgrades, prompt regarding system attach prior to upgrading UA
     packages.
-    """
-    if upgrade_packages:
-        if os.getuid() != 0:
-            print(status.MESSAGE_SECURITY_APT_NON_ROOT)
-            return
-
-        apt.run_apt_command(
-            cmd=["apt-get", "update"],
-            error_msg=status.MESSAGE_APT_UPDATE_FAILED,
-            print_cmd=True,
-        )
 
-        apt.run_apt_command(
-            cmd=["apt-get", "install", "--only-upgrade", "-y"]
-            + upgrade_packages,
-            error_msg=status.MESSAGE_APT_INSTALL_FAILED,
-            env={"DEBIAN_FRONTEND": "noninteractive"},
-            print_cmd=True,
-        )
+    :return: True if package upgrade completed or unneeded, False otherwise.
+    """
+    if not any([upgrade_packages, upgrade_packages_ua]):
+        return True
+    if os.getuid() != 0:
+        print(status.MESSAGE_SECURITY_APT_NON_ROOT)
+        return False
     if upgrade_packages_ua:
-        if cfg.is_attached:
-            print(
-                "TODO: GH: #1402: apt commands to install missing UA updates"
-            )
-        else:
-            print(status.MESSAGE_SECURITY_UPDATE_NOT_INSTALLED_SUBSCRIPTION)
-            choice = util.prompt_choices(
-                "Choose: [S]ubscribe at ubuntu.com [A]ttach"
-                " existing token [C]ancel",
-                valid_choices=["s", "a", "c"],
-            )
-            print("TODO react to subscription choice:{} GH: #".format(choice))
+        if not cfg.is_attached:
+            if not _prompt_for_attach(cfg):
+                return False  # User opted to cancel
+    packages = sorted(upgrade_packages + upgrade_packages_ua)
+    print(
```

Comment:
We are printing outside `run_apt_command` just to make use of the colorize parts, right ?

Reply:
@lucasmoura I was wondering about dropping the apt print_cmd parameter in general because we want the same type of decoration here that we have for the ua attach token output.

Per the spec, the output of the messages should default to recommended human-centered calls to `apt update` and `apt install` instead of the machine-focused calls `apt-get update` and `apt-get install` which we have to call through python code because we also provide non-interactive environment variables.

Because these commands and the printed suggestions of text are not exactly the same as what ua client is calling, I think this precludes us from printing verbatim inside the run_apt_commands the strict command being run.

Reply:
Agreed. since we are doing that distinction, I don't see any need to support the `print_cmd` anymore